### ISSUE TITLE
Do not deleted terminal buffers when the associated window is closed

### ISSUE THREAD

**Issue by languitar** [neovim/neovim#2368]
With a master build from a few days ago, terminal buffers seem to be deleted when the windows showing them is closed. This seems to contradict the usual usage patterns where buffers live until they are explicitly deleted by the user. Moreover, I'd like to be able to use a terminal session for some time, put it into the background, an reload it into a split once I need it back.

**Comment by splinterofchaos** [comment]
This should do what you want:
```vim
autocmd TermOpen * set bufhidden=hide
```
However, unless the user explicitly defines `bufhidden` for the buffers they want, or `hidden` (the global version of this option), I think it would be inconsistent to always hide terminal buffers.

When the window of a buffer not shown in any other window is closed or switches to a different buffer, the buffer can be hidden, unloaded, deleted, or wiped, however, I don't think we abide by these options correctly right now.
```vim
terminal
call feedkeys("ils\<Cr>")  " Run ls
set bufhidden=unload
edit some-other-file
```
The above script should spawn a terminal, but then immediately open `some-other-file` and the terminal buffer should be unloaded (from memory). However, if you then run the command `:ls` you will see an `h` next to the terminal buffer, meaning "loaded, but hidden". If you then run `:buffer <that number>`, the output of `ls` will still be on screen. 

Technically, when `hidden=unload`, you can't quit or edit another file without saving or forcing, so one should not be allowed to close the terminal's last window via `:edit`.

On the other hand, when instead of switching to a new file via `:edit`, one executes `:quit`, assuming there's more than one open window, `bufhidden=hide` seems to work, but `=unload` removes it from the buffer list. That makes sense because without the buffer, the terminal serves little purpose, though a little inconsistent.

So in short, the biggest bug I see is that escaping a terminal via `:edit` doesn't respect `bufhidden` when it's set to anything other than `hide`.

**Comment by justinmk** [comment]
> With a master build from a few days ago, terminal buffers seem to be deleted when the windows showing them is closed. 

@languitar I cannot reproduce this. What are the steps to reproduce this? What is `:set hidden?` ?

> So in short, the biggest bug I see is that escaping a terminal via :edit doesn't respect bufhidden when it's set to anything other than hide.

If this is the only actual bug let's change the title of this issue.

**Comment by languitar** [comment]
@justinmk seems that this behavior got fixed since I reported the issue. In a recent master version the terminal buffers won't vanish.

**Comment by splinterofchaos** [comment]
That's great. Though I still want to keep this open due to the other issues I noted.

**Comment by languitar** [comment]
Ok, now I am back to this situation. The global `hidden` option is not set. To reproduce this:

1. Start vim
1. `:e ~/.bashrc`
1. `:ls`:
   ```
:ls                                                                                                                                                                                                                                            
  1 %a   "~/.bashrc"                    line 1
   ```
1. `:term`
1. `:ls`
   ```
:ls                                                                                                                                                                                                                                            
  1 #    "~/.bashrc"                    line 1                                                                                                                                                                                                 
  2 %a-  "term://.//31919:bash"         line 66
   ```
   As expected, I am in the terminal buffer and the bashrc buffer still exists
1. `:b 1`
1. `:ls!`
   ```
:ls!                                                                                                                                                                                                                                            
  1 %a   "~/.bashrc"                    line 1
   ```
   The terminal buffers has gone, despite I didn't do anything different then with the bashrc before.

**Comment by osenvosem** [comment]
If `bufhidden` is to `hide` in init.vim the terminal buffers still close while switching. Only if it set manually, individually for each opened terminal buffer then the terminal buffers don't close. Checked on 0.2.0 and 0.2.1-dev.

It would be nice to have the opportunity to run tasks in terminal buffers in background.

**Comment by justinmk** [comment]
> It would be nice to have the opportunity to run tasks in terminal buffers in background.

You can do so if the global option `:set hidden` is enabled. 

Working on support for buffer-local `:set buffhidden` in https://github.com/neovim/neovim/pull/6280

**Comment by AttilaVM** [comment]
Hi I defined a command to open hidden terminal buffers

```
function! HiddenTerminal()
  :terminal
  set bufhidden=hide
endfunction

command! HiddenTerminal call HiddenTerminal()
nnoremap <leader>at :HiddenTerminal<CR>
```

Which works fine so far.
